File: core/sudoku_generator.py
"""SudokuGenerator — 난이도별 퍼즐 생성 (유일해 보장)."""
import random
from typing import List, Tuple

from .sudoku_solver import SudokuSolver

Grid = List[List[int]]


class SudokuGenerator:
    DIFFICULTY_HOLES = {
        "easy": 35,
        "medium": 45,
        "hard": 55,
        "expert": 62,
    }

    @staticmethod
    def _empty_grid() -> Grid:
        return [[0] * 9 for _ in range(9)]

    @classmethod
    def _fill_full(cls, grid: Grid) -> bool:
        """백트래킹으로 완성된 보드 생성."""
        for r in range(9):
            for c in range(9):
                if grid[r][c] == 0:
                    nums = list(range(1, 10))
                    random.shuffle(nums)
                    for n in nums:
                        if SudokuSolver._can_place(grid, r, c, n):
                            grid[r][c] = n
                            if cls._fill_full(grid):
                                return True
                            grid[r][c] = 0
                    return False
        return True

    @classmethod
    def generate(cls, difficulty: str = "easy") -> Tuple[Grid, Grid]:
        """(puzzle, solution) 반환. puzzle은 0이 빈칸, solution은 완성된 보드."""
        solution = cls._empty_grid()
        cls._fill_full(solution)
        puzzle = [row[:] for row in solution]
        holes_target = cls.DIFFICULTY_HOLES.get(difficulty, 35)

        cells = [(r, c) for r in range(9) for c in range(9)]
        random.shuffle(cells)

        holes = 0
        for (r, c) in cells:
            if holes >= holes_target:
                break
            backup = puzzle[r][c]
            if backup == 0:
                continue
            puzzle[r][c] = 0
            if SudokuSolver.has_unique_solution(puzzle):
                holes += 1
            else:
                puzzle[r][c] = backup
        return puzzle, solution

File: Flutter/lib/core/sudoku_generator.dart
import 'dart:math';
import 'sudoku_solver.dart';

/// 난이도별 빈 칸 수
const Map<String, int> difficultyHoles = {
  'easy': 35,
  'medium': 45,
  'hard': 55,
  'expert': 62,
};

/// 퍼즐 생성 결과 (compute()용 최상위 함수에서 사용)
class PuzzleData {
  final List<List<int>> puzzle;
  final List<List<int>> solution;
  const PuzzleData(this.puzzle, this.solution);
}

/// 수도쿠 퍼즐 생성기
/// Python sudoku_generator.py → Dart 변환
class SudokuGenerator {
  /// 난이도에 맞는 퍼즐과 정답 쌍을 반환
  static PuzzleData generate(String difficulty) {
    final solution = List.generate(9, (_) => List.filled(9, 0));
    _fillFull(solution);
    final puzzle = List.generate(9, (r) => List<int>.from(solution[r]));

    final holesTarget = difficultyHoles[difficulty] ?? 35;
    final rng = Random();

    // 모든 셀 좌표 셔플
    final cells = [for (int r = 0; r < 9; r++) for (int c = 0; c < 9; c++) (r, c)];
    cells.shuffle(rng);

    int holes = 0;
    for (final (r, c) in cells) {
      if (holes >= holesTarget) break;
      final backup = puzzle[r][c];
      if (backup == 0) continue;
      puzzle[r][c] = 0;
      if (SudokuSolver.hasUniqueSolution(puzzle)) {
        holes++;
      } else {
        puzzle[r][c] = backup;
      }
    }

    return PuzzleData(puzzle, solution);
  }

  static bool _fillFull(List<List<int>> grid) {
    final rng = Random();
    for (int r = 0; r < 9; r++) {
      for (int c = 0; c < 9; c++) {
        if (grid[r][c] == 0) {
          final nums = List.generate(9, (i) => i + 1)..shuffle(rng);
          for (final n in nums) {
            if (SudokuSolver.canPlace(grid, r, c, n)) {
              grid[r][c] = n;
              if (_fillFull(grid)) return true;
              grid[r][c] = 0;
            }
          }
          return false;
        }
      }
    }
    return true;
  }
}

/// compute() isolate 진입점 (최상위 함수여야 함)
PuzzleData generatePuzzleIsolate(String difficulty) {
  return SudokuGenerator.generate(difficulty);
}
